packages/firestore/lite/src/api/reference.ts

```diff
@@ -49,3 +55,222 @@ export class DocumentReference<T = firestore.DocumentData>
+    if (DocumentKey.isDocumentKey(path)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${path}). Path points to a document.`
+      );
+    }
+    return new CollectionReference(parent, path);
+  } else {
+    const doc = tryCast(parent, DocumentReference);
+    const absolutePath = doc._key.path.child(path);
+    if (DocumentKey.isDocumentKey(absolutePath)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${absolutePath}). Path points to a document.`
+      );
+    }
+    return new CollectionReference(doc.firestore, absolutePath);
+  }
+}
+
+export function doc(
+  firestore: FirebaseFirestore,
+  documentPath: string
+): DocumentReference<firestore.DocumentData>;
+export function doc<T>(
+  reference: CollectionReference<T>,
+  documentPath?: string
+): DocumentReference<T>;
+export function doc<T>(
+  parent: firestore.FirebaseFirestore | firestore.CollectionReference<T>,
+  relativePath?: string
+): DocumentReference {
+  // We allow omission of 'pathString' but explicitly prohibit passing in both
+  // 'undefined' and 'null'.
+  if (arguments.length === 1) {
+    relativePath = AutoId.newId();
+  }
+
+  if (!relativePath) {
+    throw new FirestoreError(
+      Code.INVALID_ARGUMENT,
+      `Invalid path (${relativePath}). Empty paths are not supported.`
+    );
+  }
+
+  const path = ResourcePath.fromString(relativePath);
+  if (parent instanceof Firestore) {
+    if (!DocumentKey.isDocumentKey(path)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${path}). Path points to a collection.`
+      );
+    }
+    return new DocumentReference(parent, new DocumentKey(path));
+  } else {
+    const coll = tryCast(parent, CollectionReference);
+    const absolutePath = coll._path.child(path);
+    if (!DocumentKey.isDocumentKey(absolutePath)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${absolutePath}). Path points to a collection.`
+      );
+    }
+    return new DocumentReference(
+      coll.firestore,
+      new DocumentKey(absolutePath),
+      coll._converter
+    );
+  }
+}
+
+export function parent(
+  reference: CollectionReference<unknown>
+): DocumentReference<firestore.DocumentData> | null;
+export function parent<T>(
+  reference: DocumentReference<T>
+): CollectionReference<T>;
+export function parent<T>(
+  child: firestore.CollectionReference<unknown> | firestore.DocumentReference<T>
+): DocumentReference<firestore.DocumentData> | CollectionReference<T> | null {
+  if (child instanceof CollectionReference) {
+    const parentPath = child._path.popLast();
+    if (parentPath.isEmpty()) {
+      return null;
+    } else {
+      return new DocumentReference(
+        child.firestore,
+        new DocumentKey(parentPath)
+      );
+    }
+  } else {
+    const doc = tryCast(child, DocumentReference) as DocumentReference<T>;
```

Comment:
Suggestion: Instead of casting the return value with the `as` keyword consider passing it as the generic argument on the function. The net result is often the same but the former conveys the return value "should have members from this type" vs "is of this type" and I believe you're trying to communicate the latter.

Reply:
I wasn't able to figure out how to do this for a type that is already generic. The function signature is already:

```
export function cast<T>(
  obj: object,
  constructor: { new (...args: any[]): T }
): T {
```

Changing it to accept two generic arguments results in a compile error (type `T` is not generic).

I was able to improve this line a little bit by hard-coding T:

```
const doc = cast<DocumentReference<T>>(child, DocumentReference);
```

Reply:
What you changed it to is exactly what I was looking for. I should have been more explicit.